### ISSUE TITLE
Email Confirmation for AFFINE Cloud SelfHosted ?​

### ISSUE THREAD

**Issue by luckylinux** [toeverything/AFFiNE#6569]
### What happened?

Excuse me if it's something obvious, but ...

According to https://docs.affine.pro/docs/self-host-affine it just says to Sign In with the Configured Email and Password.

I had this configured in `compose.yml` in the `environment:` section:
```
      - AFFINE_ADMIN_EMAIL=admin@MYDOMAIN.TLD
      - AFFINE_ADMIN_PASSWORD=XXXXXXXXXXXXXXXXXx
```

The Documentation doesn't mention anything about the Requirement of setting up the SMTP Email Settings so that Affine can send a confirmation Email ...

I get this message in the Browser (Firefox):
```
Send email error
Please back to home and try again
```

I am running Affine in a Local Environment (Local Homelab) but with Traefik Proxy in front of it.

Maybe the problem arises by the fact that it is detecting a valid Domain (I used that to have Letsencrypt Validated SSL Certificates) and NOT and IP Address / `localhost` ?

Unsure ...

### Distribution version

Linux

### What browsers are you seeing the problem on if you're using web version?

- Firefox
- Chromium 
### Are you self-hosting?

- [X] Yes

### Relevant log output

```shell
Logs of `affine-server` container:

running predeploy script.
yarn run v1.22.19
$ yarn prisma migrate deploy && node --import ./scripts/register.js ./dist/data/index.js run
$ /app/node_modules/.bin/prisma migrate deploy
Prisma schema loaded from schema.prisma
Datasource "db": PostgreSQL database "affine", schema "public" at "affine-postgresql:5432"

26 migrations found in prisma/migrations


No pending migrations to apply.
(node:89) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [NestFactory] Starting Nest application...
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] CommandRootModule dependencies initialized +31ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] CacheModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] PrismaModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] RateLimiterModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] SyncModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] ServerConfigModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] WebSocketModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] GqlModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] DiscoveryModule dependencies initialized +2ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] DiscoveryModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] EventModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] MutexModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] MetricsModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] MailerModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] StorageProviderModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] HelpersModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] ServeStaticModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] EventEmitterModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] ScheduleModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] CommandRunnerModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] FeatureModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] StorageModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] MailModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] DocModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] ThrottlerModule dependencies initialized +4ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] CliAppModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] QuotaModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] GraphQLSchemaBuilderModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] EventsModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] WorkspaceModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] GraphQLModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] UserModule dependencies initialized +0ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [InstanceLoader] AuthModule dependencies initialized +1ms
[Nest] 89  - 04/15/2024, 5:02:57 PM     LOG [RunCommand] Done 0 migrations
Done in 6.24s.
(node:112) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [NestFactory] Starting Nest application...
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] ConfigModule dependencies initialized +29ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] CacheModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] PrismaModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] RateLimiterModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] SyncModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] ServerConfigModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] WebSocketModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] GqlModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] DiscoveryModule dependencies initialized +8ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] MetricsModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] MailerModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] StorageProviderModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] HelpersModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] EventModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] MutexModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] ServeStaticModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] AppModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] EventEmitterModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] ScheduleModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] MailModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] FeatureModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] StorageModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] DocModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] ThrottlerModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] QuotaModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] GraphQLSchemaBuilderModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] EventsModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] WorkspaceModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] GraphQLModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] UserModule dependencies initialized +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [InstanceLoader] AuthModule dependencies initialized +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "client-handshake-sync" message +75ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "client-handshake-awareness" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "client-leave-sync" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "client-leave-awareness" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "client-pre-sync" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "client-update-v2" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "doc-load-v2" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "awareness-init" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [WebSocketsController] EventsGateway subscribed to the "awareness-update" message +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RoutesResolver] AuthController {/api/auth}: +2ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/auth/sign-in, POST} route +2ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/auth/sign-out, GET} route +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/auth/magic-link, GET} route +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/auth/session, GET} route +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/auth/sessions, GET} route +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/auth/challenge, GET} route +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RoutesResolver] UserAvatarController {/api/avatars}: +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/avatars/:id, GET} route +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RoutesResolver] WorkspacesController {/api/workspaces}: +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/workspaces/:id/blobs/:name, GET} route +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/workspaces/:id/docs/:guid, GET} route +0ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [RouterExplorer] Mapped {/api/workspaces/:id/docs/:guid/histories/:timestamp, GET} route +1ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [DocManager] Use Database
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [DocManager] Automation started
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [GraphQLModule] Mapped {/graphql, POST} route +119ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [NestApplication] Nest application successfully started +50ms
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [App] AFFiNE Server is running in [selfhosted] mode
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [App] Listening on http://0.0.0.0:3010
[Nest] 112  - 04/15/2024, 5:03:00 PM     LOG [App] And the public server should be recognized as http://affine.MYDOMAIN.TLD
[Nest] 112  - 04/15/2024, 5:03:34 PM   ERROR [ExceptionsHandler] Mailer service is not configured.
Error: Mailer service is not configured.
    at MailService.sendMail (file:///app/dist/fundamentals/mailer/mail.service.js:26:19)
    at MailService.sendSignUpMail (file:///app/dist/fundamentals/mailer/mail.service.js:86:21)
    at AuthService.sendSignInEmail (file:///app/dist/core/auth/service.js:311:33)
    at AuthController.sendSignInEmail (file:///app/dist/core/auth/controller.js:61:40)
    at async AuthController.signIn (file:///app/dist/core/auth/controller.js:45:28)
[Nest] 112  - 04/15/2024, 5:10:36 PM   ERROR [ExceptionsHandler] Mailer service is not configured.
Error: Mailer service is not configured.
    at MailService.sendMail (file:///app/dist/fundamentals/mailer/mail.service.js:26:19)
    at MailService.sendSignUpMail (file:///app/dist/fundamentals/mailer/mail.service.js:86:21)
    at AuthService.sendSignInEmail (file:///app/dist/core/auth/service.js:311:33)
    at AuthController.sendSignInEmail (file:///app/dist/core/auth/controller.js:61:40)
    at async AuthController.signIn (file:///app/dist/core/auth/controller.js:45:28)
[Nest] 112  - 04/15/2024, 5:11:46 PM   ERROR [ExceptionsHandler] Mailer service is not configured.
Error: Mailer service is not configured.
    at MailService.sendMail (file:///app/dist/fundamentals/mailer/mail.service.js:26:19)
    at MailService.sendSignUpMail (file:///app/dist/fundamentals/mailer/mail.service.js:86:21)
    at AuthService.sendSignInEmail (file:///app/dist/core/auth/service.js:311:33)
    at AuthController.sendSignInEmail (file:///app/dist/core/auth/controller.js:61:40)
    at async AuthController.signIn (file:///app/dist/core/auth/controller.js:45:28)

```
```


### Anything else?

Distribution Host: Debian Bookworm 12 GNU/Linux
Architecture Host: amd64
Podman: 4.9.3 (Alternative to Docker)
Traefik: 2.10 (2.11 isn't playing nice)
Affine: ghcr.io/toeverything/affine-graphql:stable

**Comment by JimmFly** [comment]
We are still discussing solutions to the problem of selfhost not being able to send emails. If you have any ideas, please feel free to create a PR.

**Comment by JimmFly** [comment]
Maybe you can discuss it in this [discussions](https://github.com/toeverything/AFFiNE/discussions/5975)

**Comment by EYHN** [comment]
If you log in using the `AFFINE_ADMIN_EMAIL` account, you will not need to receive emails.

**Comment by EYHN** [comment]
There are many similar situations in discussion, please discuss in discussion

**Comment by luckylinux** [comment]
@EYHN As I said I tried to login with the `AFFINE_ADMIN_EMAIL` and it still wants to submit an Email.

**Comment by bigm0re** [comment]
> @EYHN As I said I tried to login with the `AFFINE_ADMIN_EMAIL` and it still wants to submit an Email.

me 2,any solution?

**Comment by lakisan1** [comment]
> @EYHN As I said I tried to login with the `AFFINE_ADMIN_EMAIL` and it still wants to submit an Email.

Did you manage to solve the login problem? I am in the same spot :(

**Comment by luckylinux** [comment]
Nope, gave up on this, have far too much other stuff going on at the time. 

I'd say the developers need to fix this ... Otherwise you could setup some dummy system like a Docker/Podman Container with `yoryan/mailrise` (tested) / `axllent/mailpit` (tested) / `rnwood/smtp4dev` (I didn't use it, but I guess similar to `axllent/mailpit`) Images, send yourself an email and "fake it" like that ...

**Comment by Dtanguy** [comment]
Exact same problem here

**Comment by bmcgonag** [comment]
Just an FYI - I was able to get around this by going into the .affine/self-host/config directory (created after running initial `docker-compose up`), and then editing the affine.env.js file.  The values for Email are listed as

    MAILER_HOST: 'smtp.domain.com',
    MAILER_PORT: ['587', 'int'],
    MAILER_USER: 'your-email-user@domain.com',
    MAILER_PASSWORD: 'your-email-user-password-here',
    MAILER_SENDER: 'your-email-user@domain.com',
    MAILER_SECURE: ['true', 'boolean'],

I hope this helps someone else get things going.

**Comment by alardus** [comment]
Have done email sending to work as follow:

1. Do not make any changes in the affine.env.js file
2. Make sure the following variables are set in the compose.yml file at the environment section:
```
- MAILER_HOST=smtp.domain.com
- MAILER_PORT=587
- MAILER_USER=your_user@domain.com
- MAILER_PASSWORD=your_user_password
- MAILER_SENDER=your_user@domain.com
```
Be sure to use the app password, not the normal one.
3. Recreate the container with the command `compose up --build --force-recreate -d`

**Comment by merrime-n** [comment]
@alardus Thanks. Your solution did work like a charm. I suggest the OP to mark your comment as the solution.

**Comment by luckylinux** [comment]
How is that a Solution ?

The Point is that we shouldn't need to provide & get an Email Confirmation for a single User & single Instance.

I don't think I have another self-hosted Application that requires this to be honest.

Yes, @alardus Procedure together with something like `mailpit` / `mailsandbox` Server would work, but it's still an extra Procedure which shouldn't be required to begin with in self-hosted Environments IMHO.

**Comment by merrime-n** [comment]
@luckylinux  I agree completely—the cleanest approach would be to let the admin sign in with the credentials already defined in `.env`, no e-mail involved. Unfortunately the app won’t allow that yet, and the SMTP section of the docs is still thin.  

Until native password-only login lands, the most workable stop-gap is to set the SMTP variables in your environment (and point them at something lightweight like Mailpit if you don’t want to run a full mail server). It’s an extra step, but right now it’s the only way to satisfy the confirmation check without patching the code.

**Comment by Billiam** [comment]
If you don't need to worry about registration, you can also manually verify users by connecting directly to the database and updating their verified status:

```sql
UPDATE users SET email_verified = NOW()
```

**Comment by CKeckNorthIO** [comment]
Unfortunately, this is still no complete solution. Even if you set "MAILER_SECURE" to "false", the connection is still made with STARTTLS, which fails if the remote server does not support this:

Sep 26 07:31:42 xxxx postfix/smtpd[994421]: connect from affine.local [192.168.10.136]
Sep 26 07:31:42 xxxx postfix/smtpd[994421]: SSL_accept error from affine.local[192.168.10.136]: lost connection
Sep 26 07:31:42 xxxx postfix/smtpd[994421]: lost connection after STARTTLS from affine.local[192.168.10.136]
Sep 26 07:31:42 xxxx postfix/smtpd[994421]: disconnect from affine.local[192.168.10.136] ehlo=1 starttls=0/1 commands=1/2

Please provide a way to allow an unencrypted connection, too. "MAILER_SECURE" does not work somehow.

**Comment by Snuupy** [comment]
> UPDATE users SET email_verified = NOW()

close, I needed `UPDATE public.users SET email_verified = NOW();`

**Comment by MrZoule** [comment]
Hello,
We have a delivery issue too. We made a lot of test and discovered finally the problem in the SMTP logs:
```
[2025-10-15 16:49:58] DEBUG [#zt4rxh3xpfc3c7iz] C: EHLO AFFiNE Server
[2025-10-15 16:49:58] DEBUG [#zt4rxh3xpfc3c7iz] S: 501 Error: syntax: EHLO hostname
```
A MAILER_HOSTNAME env variable should be available to announce a correct EHLO value, no?

**Comment by MrZoule** [comment]
Hmm my bad, resolved by https://github.com/toeverything/AFFiNE/pull/13632
But documentation is not up to date.

**Comment by nolazelaya** [comment]
> Hmm my bad, resolved by [#13632](https://github.com/toeverything/AFFiNE/pull/13632) But documentation is not up to date.

Agreed, looks like there's a [request to update docs](https://github.com/toeverything/AFFiNE/discussions/13853)
